### PR TITLE
fix: resolve signed integer overflow UB in CoinJoin priority and timeout

### PR DESCRIPTION
**Pull request by thepastaclaw**
## Summary

Fix two signed integer overflow UB issues in CoinJoin code, found during fuzz testing.

### `CalculateAmountPriority` (common.h)

The return type is `int` but the computation `-(nInputAmount / COIN)` operates on
`int64_t` values. When `nInputAmount` is extremely large (e.g. near `MAX_MONEY`),
the result exceeds `INT_MAX` and the implicit narrowing to `int` is undefined
behavior under UBSan.

**Fix:** Clamp the `int64_t` result to `[INT_MIN, INT_MAX]` before returning.
This preserves the existing sort ordering for all realistic inputs while making
extreme values well-defined.

### `IsTimeOutOfBounds` (coinjoin.cpp)

The expression `current_time - nTime` overflows when the two `int64_t` values
differ by more than `INT64_MAX` (e.g. one large positive, one large negative).

**Fix:** Compute the absolute difference using unsigned arithmetic, which is
well-defined for all inputs.

## Validation

- Both functions are non-consensus (CoinJoin sort priority and queue timeout only)
- Neither overflow is exploitable — CoinJoin queue entries require valid MN signatures,
  and the priority function only affects local sort order
- The fixes preserve identical behavior for all realistic inputs
- Found via UBSan-instrumented fuzz testing on the `ci/fuzz-regression` branch

### CODE REVIEW

**Comment by thepastaclaw** [comment]
@coderabbitai review

File: src/coinjoin/coinjoin.cpp [hunk 1]
```diff
@@ -57,8 +57,9 @@ bool CCoinJoinQueue::CheckSignature(const CBLSPublicKey& blsPubKey) const
 
 bool CCoinJoinQueue::IsTimeOutOfBounds(int64_t current_time) const
 {
-    return current_time - nTime > COINJOIN_QUEUE_TIMEOUT ||
-           nTime - current_time > COINJOIN_QUEUE_TIMEOUT;
+    const uint64_t diff = (current_time > nTime) ? static_cast<uint64_t>(current_time) - static_cast<uint64_t>(nTime)
+                                                 : static_cast<uint64_t>(nTime) - static_cast<uint64_t>(current_time);
+    return diff > static_cast<uint64_t>(COINJOIN_QUEUE_TIMEOUT);
```
Comment (thepastaclaw) [inline src/coinjoin/coinjoin.cpp]: **🟡 Suggestion: No regression test for extreme timestamp values in IsTimeOutOfBounds**

The fix is correct: unsigned subtraction after a signed comparison produces the exact mathematical absolute difference for all int64_t pairs, including mixed-sign and extreme values. However, existing tests in coinjoin_queue_tests.cpp (deltas of ±15, ±60) and coinjoin_inouts_tests.cpp (delta of COINJOIN_QUEUE_TIMEOUT+1) only cover small time deltas around realistic timestamps. No test exercises the scenario that triggered the UBSan finding — e.g., current_time = INT64_MAX and nTime = INT64_MIN, or vice versa. A targeted test would prevent regressions and document the fix motivation.

<sub>source: ['claude', 'codex']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/coinjoin/coinjoin.cpp`:
- [SUGGESTION] lines 60-62: No regression test for extreme timestamp values in IsTimeOutOfBounds
  The fix is correct: unsigned subtraction after a signed comparison produces the exact mathematical absolute difference for all int64_t pairs, including mixed-sign and extreme values. However, existing tests in coinjoin_queue_tests.cpp (deltas of ±15, ±60) and coinjoin_inouts_tests.cpp (delta of COINJOIN_QUEUE_TIMEOUT+1) only cover small time deltas around realistic timestamps. No test exercises the scenario that triggered the UBSan finding — e.g., current_time = INT64_MAX and nTime = INT64_MIN, or vice versa. A targeted test would prevent regressions and document the fix motivation.
```
</details>

File: src/coinjoin/common.h [hunk 2]
```diff
@@ -127,7 +129,8 @@ constexpr int CalculateAmountPriority(CAmount nInputAmount)
     }
 
     //nondenom return largest first
-    return -1 * (nInputAmount / COIN);
+    const int64_t val = -(nInputAmount / COIN);
+    return int(std::clamp<int64_t>(val, std::numeric_limits<int>::min(), std::numeric_limits<int>::max()));
```
Comment (thepastaclaw) [inline src/coinjoin/common.h]: **🟡 Suggestion: No regression test for CalculateAmountPriority clamp boundary**

The clamp is correct. Note that with valid money amounts, the old code would not have overflowed int: -(MAX_MONEY / COIN) = -21000000, well within INT_MIN..INT_MAX. The UB requires nInputAmount > INT_MAX * COIN (~2.1×10¹⁷), which exceeds MAX_MONEY but is representable in int64_t. Since CalculateAmountPriority accepts CAmount (int64_t) without a MoneyRange guard, fuzz testing can reach these values. There are no tests for this function at all — the function is constexpr, so a static_assert or compile-time test could verify the clamping directly.

<sub>source: ['claude', 'codex']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/coinjoin/common.h`:
- [SUGGESTION] lines 132-133: No regression test for CalculateAmountPriority clamp boundary
  The clamp is correct. Note that with valid money amounts, the old code would not have overflowed int: -(MAX_MONEY / COIN) = -21000000, well within INT_MIN..INT_MAX. The UB requires nInputAmount > INT_MAX * COIN (~2.1×10¹⁷), which exceeds MAX_MONEY but is representable in int64_t. Since CalculateAmountPriority accepts CAmount (int64_t) without a MoneyRange guard, fuzz testing can reach these values. There are no tests for this function at all — the function is constexpr, so a static_assert or compile-time test could verify the clamping directly.
```
</details>

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

Clean, correct two-line fix for signed integer overflow UB in CoinJoin timeout and priority. Both arithmetic changes are mathematically sound and preserve original semantics. The only gap is the absence of regression tests exercising the extreme-value inputs that motivated these fixes.

*Reviewed commit: ab4bea6f*

🟡 2 suggestion(s) | 💬 1 nitpick(s)

### 1 additional finding

<details>
<summary>💬 <b>nitpick:</b> Pre-existing: float-to-int implicit narrowing on denomination priority return</summary>

`src/coinjoin/common.h (line 125)`

Line 125 returns `(float)COIN / *optDenom * 10000`, which undergoes implicit float-to-int conversion. With current denominations the maximum result is ~10,000,000 (for the smallest denomination COIN/1000+1), safely within INT_MAX. But this is the same class of narrowing issue this PR fixes for the nondenom path. Not blocking since current values are safe, but worth noting for consistency.

</details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/coinjoin/coinjoin.cpp`:
- [SUGGESTION] lines 60-62: No regression test for extreme timestamp values in IsTimeOutOfBounds
  The fix is correct: unsigned subtraction after a signed comparison produces the exact mathematical absolute difference for all int64_t pairs, including mixed-sign and extreme values. However, existing tests in coinjoin_queue_tests.cpp (deltas of ±15, ±60) and coinjoin_inouts_tests.cpp (delta of COINJOIN_QUEUE_TIMEOUT+1) only cover small time deltas around realistic timestamps. No test exercises the scenario that triggered the UBSan finding — e.g., current_time = INT64_MAX and nTime = INT64_MIN, or vice versa. A targeted test would prevent regressions and document the fix motivation.

In `src/coinjoin/common.h`:
- [SUGGESTION] lines 132-133: No regression test for CalculateAmountPriority clamp boundary
  The clamp is correct. Note that with valid money amounts, the old code would not have overflowed int: -(MAX_MONEY / COIN) = -21000000, well within INT_MIN..INT_MAX. The UB requires nInputAmount > INT_MAX * COIN (~2.1×10¹⁷), which exceeds MAX_MONEY but is representable in int64_t. Since CalculateAmountPriority accepts CAmount (int64_t) without a MoneyRange guard, fuzz testing can reach these values. There are no tests for this function at all — the function is constexpr, so a static_assert or compile-time test could verify the clamping directly.
```
</details>

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

Test-only follow-up commit adding comprehensive regression tests for both UB fixes. All prior review suggestions are now addressed: `queue_timestamp_extreme_values` covers INT64_MIN/INT64_MAX edge cases for `IsTimeOutOfBounds`, and `calculate_amount_priority_clamp` (plus two `static_assert`s) verifies the clamping boundary for `CalculateAmountPriority`. Tests are well-structured and exercise the exact scenarios that motivated the fixes.

*Reviewed commit: 1053f373*

**Comment by thepastaclaw** [comment]
Updated per knst's feedback: replaced magic numbers in tests with symbolic expressions (`-(MAX_MONEY / COIN)`, `-INT_MAX`) so the expected values are self-documenting.

**Review by thepastaclaw — COMMENTED** [review]
**Incremental Review** (push `4cc88eaf` vs prior reviewed `1053f373`)

**Delta:** 1 file changed — `src/test/coinjoin_queue_tests.cpp` (4 insertions, 5 deletions)

Test-only commit replacing hardcoded magic numbers with symbolic constants: `-21000000` → `-(MAX_MONEY / COIN)`, `-2147483647` → `-INT_MAX`. Pure readability improvement, no behavioral change.

Prior review stands — no new findings. ✅

File: src/coinjoin/common.h (hunk 2 above)
Comment (knst) [reply src/coinjoin/common.h]: The value for this function is coming as effective_value:
```
    /** The value of the UTXOs after deducting the cost of spending them at the effective feerate. */
    CAmount effective_value{0};
```
which is never bigger than MAX_MONEY.

This fix is unexpected complication of code IMO

File: src/test/coinjoin_queue_tests.cpp [hunk 3]
```diff
@@ -96,4 +101,47 @@ BOOST_AUTO_TEST_CASE(queue_timestamp_validation)
...
+    // Zero vs extreme values → huge gap → out of bounds
+    q.nTime = 0;
+    BOOST_CHECK(q.IsTimeOutOfBounds(INT64_MAX));
+
+    q.nTime = 0;
+    BOOST_CHECK(q.IsTimeOutOfBounds(INT64_MIN));
+}
+
+static_assert(CoinJoin::CalculateAmountPriority(MAX_MONEY) == -(MAX_MONEY / COIN));
+static_assert(CoinJoin::CalculateAmountPriority(static_cast<CAmount>(INT64_MAX)) == std::numeric_limits<int>::min());
```
Comment (knst) [inline src/test/coinjoin_queue_tests.cpp]: this unit test doesn't make any sense, because CAmount for CJ never could be more than MAX_MONEY, isn't it?

File: src/coinjoin/common.h (hunk 2 above)
Comment (thepastaclaw) [reply src/coinjoin/common.h]: You're right that in practice `effective_value` is always bounded by `MAX_MONEY`, so `-(MAX_MONEY / COIN) = -21000000` fits comfortably in `int`. The narrowing is safe for all realistic inputs.

The clamp was added because UBSan flagged this during fuzz testing — the fuzz harness can feed arbitrary `CAmount` values to `CalculateAmountPriority` without going through the wallet code path that enforces the money range constraint. The implicit `int64_t → int` narrowing is technically undefined behavior per the standard when the value doesn't fit, even if it's unreachable in production.

That said, I can see the argument that defending against inputs that can't occur in practice adds unnecessary complexity. A few options:

1. **Keep the clamp** (current) — zero-cost for realistic values, makes the function correct for all inputs
2. **Replace with `Assume()`** — e.g. `Assume(std::abs(val) <= INT_MAX)` to document the precondition without the clamp
3. **Remove the clamp entirely** — just add a comment that narrowing is safe because callers always pass values ≤ MAX_MONEY

Happy to go with whichever you prefer. The `IsTimeOutOfBounds` fix (unsigned arithmetic) is the more critical one — it can trigger on any signed subtraction with large deltas.

File: src/test/coinjoin_queue_tests.cpp (hunk 3 above)
Comment (thepastaclaw) [reply src/test/coinjoin_queue_tests.cpp]: Agreed — the `calculate_amount_priority_clamp` test and `static_assert` are only meaningful if we keep the clamp. If we drop the clamp per the discussion above, these tests should go too.

The `queue_timestamp_extreme_values` tests (for `IsTimeOutOfBounds`) are still important though — those exercise the actual UB that existed in the signed subtraction path.

**Comment by UdjinM6** [comment]
We could simply bail out on invalid values instead of trying to handle them i.e. `if (current_time < 0 || nTime < 0) return true;` in `IsTimeOutOfBounds()` and `if (!MoneyRange(nInputAmount)) return 0;` in `CalculateAmountPriority()`

**Comment by thepastaclaw** [comment]
Done — updated in 9f2f390. Both functions now bail out on invalid inputs with early-return guards:

- `IsTimeOutOfBounds`: `if (current_time < 0 || nTime < 0) return true;` — original subtraction logic preserved
- `CalculateAmountPriority`: `if (nInputAmount < 0 || nInputAmount > MAX_MONEY) return 0;` — inlined the MoneyRange check to keep the function constexpr

Also dropped the ci-retrigger commit and updated the test expectations to match.

File: src/coinjoin/common.h (hunk 2 above)
Comment (thepastaclaw) [reply src/coinjoin/common.h]: Updated per UdjinM6's suggestion — replaced the clamp with a simple MoneyRange guard: `if (nInputAmount < 0 || nInputAmount > MAX_MONEY) return 0;`. The original `return -1 * (nInputAmount / COIN)` is preserved for valid inputs.

File: src/test/coinjoin_queue_tests.cpp (hunk 3 above)
Comment (thepastaclaw) [reply src/test/coinjoin_queue_tests.cpp]: Updated — the test now verifies the MoneyRange guard behavior: out-of-range amounts (`INT64_MAX`, `-1`, `MAX_MONEY + 1`) all return 0. The `static_assert` for `INT64_MAX` now expects 0 instead of `INT_MIN`.

File: src/coinjoin/common.h [hunk 4]
```diff
@@ -127,6 +127,7 @@ constexpr int CalculateAmountPriority(CAmount nInputAmount)
     }
 
     //nondenom return largest first
+    if (nInputAmount < 0 || nInputAmount > MAX_MONEY) return 0;
```
[the remark this answers is not in the document]
Comment (thepastaclaw) [reply src/coinjoin/common.h]: Good catch — moved the MoneyRange guard to the top of the function so it's checked before `nInputAmount < COIN`. Added a `static_assert` for `-1` as well.

**Review by PastaPastaPasta — APPROVED** [review]
utACK e8ec63a1e090f4e6bcdfa3cd48891c8672b51433

**Review by UdjinM6 — APPROVED** [review]
utACK e8ec63a1e090f4e6bcdfa3cd48891c8672b51433

**Review by PastaPastaPasta — APPROVED** [review]
utACK e8ec63a1e090f4e6bcdfa3cd48891c8672b51433
